File: test-suite/rust/clean/header_injection.rs
use axum::extract::Query;
use http::HeaderValue;
use std::collections::HashMap;

struct Response;
struct ResponseBuilder;

impl Response {
    fn builder() -> ResponseBuilder {
        ResponseBuilder
    }
}

impl ResponseBuilder {
    fn header(self, _name: &str, _value: impl AsRef<str>) -> Self {
        self
    }
}

fn safe_header_value(raw: &str) -> String {
    raw.replace('\r', "").replace('\n', "")
}

fn query_value_in_safe_header(Query(params): Query<HashMap<String, String>>) -> ResponseBuilder {
    let display_name = safe_header_value(params.get("name").map(String::as_str).unwrap_or_default());
    Response::builder().header("X-Display-Name", display_name)
}

fn encoded_filename(Query(params): Query<HashMap<String, String>>) -> ResponseBuilder {
    let filename = params.get("filename").map(String::as_str).unwrap_or_default();
    let encoded = urlencoding::encode(filename);
    Response::builder().header("Content-Disposition", format!("attachment; filename={encoded}"))
}

fn header_value_constructor(Query(params): Query<HashMap<String, String>>) -> Result<ResponseBuilder, &'static str> {
    let trace_id = params.get("trace_id").map(String::as_str).unwrap_or_default();
    let header = HeaderValue::from_str(trace_id).map_err(|_| "invalid header")?;
    Ok(Response::builder().header("X-Upstream-Trace", header.to_str().unwrap_or_default()))
}

fn reject_crlf_before_header(Query(params): Query<HashMap<String, String>>) -> Result<ResponseBuilder, &'static str> {
    let reason = params.get("reason").cloned().unwrap_or_default();
    if reason.contains('\r') || reason.contains('\n') {
        return Err("invalid header");
    }

    Ok(Response::builder().header("X-Return-Reason", reason))
}

File: test-suite/elixir/clean/header_injection.ex
defmodule CleanHeaderInjection do
  import Plug.Conn

  @allowed_redirect_hosts MapSet.new(["app.example.com", "accounts.example.com"])

  def safe_header_value(raw) do
    raw
    |> String.replace("\r", "")
    |> String.replace("\n", "")
  end

  def encoded_filename(raw) do
    URI.encode(raw)
  end

  def reject_crlf(value) do
    if String.contains?(value, "\r") or String.contains?(value, "\n") do
      raise ArgumentError, "invalid header"
    end

    value
  end

  def safe_redirect_url(raw) do
    if String.starts_with?(raw, "/") and not String.starts_with?(raw, "//") do
      raw
    else
      uri = URI.parse(raw)

      unless uri.scheme == "https" and MapSet.member?(@allowed_redirect_hosts, uri.host) do
        raise ArgumentError, "blocked redirect"
      end

      raw
    end
  end

  def display_name(conn) do
    name = safe_header_value(conn.params["display_name"] || "guest")

    conn
    |> put_resp_header("x-display-name", name)
    |> send_resp(200, "ok")
  end

  def download(conn) do
    filename = encoded_filename(Map.get(conn.query_params, "filename") || "report.csv")

    conn
    |> put_resp_header("content-disposition", "attachment; filename=\"#{filename}\"")
    |> send_resp(200, "ok")
  end

  def trace_header(conn) do
    trace =
      conn
      |> Plug.Conn.get_req_header("x-trace-id")
      |> List.first()
      |> safe_header_value()

    Plug.Conn.put_resp_header(conn, "x-upstream-trace", trace)
  end

  def cookie_header(conn) do
    export_name = safe_header_value(conn.cookies["export_name"] || "report")
    put_resp_header(conn, "x-export-name", export_name)
  end

  def content_type(conn) do
    media_type = safe_header_value(conn.params["content_type"] || "text/plain")
    put_resp_content_type(conn, media_type)
  end

  def rejected_customer_header(conn) do
    value = conn.params["customer"] |> reject_crlf()
    put_resp_header(conn, "x-customer", value)
  end

  def location_only(conn) do
    target = safe_redirect_url(conn.params["location"] || "/")

    conn
    |> put_resp_header("location", target)
    |> send_resp(302, "")
  end
end
